sound/soc/sof/intel/hda.c

```diff
@@ -492,10 +492,15 @@ static int hda_init(struct snd_sof_dev *sdev)
 	/* HDA base */
 	sdev->bar[HDA_DSP_HDA_BAR] = bus->remap_addr;
 
-	/* init i915 and HDMI codecs */
-	ret = hda_codec_i915_init(sdev);
-	if (ret < 0)
-		dev_warn(sdev->dev, "init of i915 and HDMI codec failed\n");
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	/* do not init i915 codec if codec is masked */
+	if (HDA_IDISP_CODEC(hda_codec_mask)) {
+		/* init i915 and HDMI codecs */
+		ret = hda_codec_i915_init(sdev);
+		if (ret < 0)
+			dev_warn(sdev->dev, "init of i915 and HDMI codec failed\n");
+	}
+#endif
```

Comment:
The ifdef was left out on purpose before, but I guess now we have to add it back as the params is not defined otherwise.

This relies on the assumption the parameter is initialized to -1 if not set by the user, but it does work now and maybe it's reasonable assumption also in the future.

Another small code cleanliness issue is that now we have a unmatched call to i915_exit(). The implementation can handle this, but it looks a bit weird. Maybe better still like this. Added the same ifdefs and mask checks to a many places in error recovery path looks pretty ugly as well.

Reply:
can't we hide the ifdefs inside of hda_codec_i915_init/exit, that way the error path is identical but we have a different internal behavior?

Reply:
I think we'd better keep this here, we may need some clean up for nocodec mode.
When SOF is in nocodec mode, do we still need to `hda_codec_i915_init` call here? I found this issue when testing this patch with nocodec. It seems even if we do not use HDA at all with nocodec mode, we will get 60s delay

Reply:
How about moving the `if (HDA_IDISP_CODEC(hda_codec_mask))` check inside `hda_codec_i915_init()`? No ifdef needed then. And no unmatched call to `hda_codec_i915_exit()` but maybe also add an immediate return from there too if no codec is detected`